### DISCUSSION TITLE
Order of fields in form

### DISCUSSION THREAD

**Discussion by yellowbooth** [Laravel-Backpack/community-forum#718] [Q&A (Help)]
I have just started trying backpack and think it's fantastic!! The only issue I am finding (which may well be my fault) is as follows:

I have added a table based select in my DealerCrudController.php
   CRUD::field('country_id')->type('select')->entity('country')->attribute('name')->model('App\Models\Country');

Problem - it appears as the first field in the forms, when I want it to appear last!

So my questions are:
1) What controls the position in the form.
2) Is there any way I can control the order of the fields?

Thanks in advance :)

**Comment by MaraIGI** [comment]
Hi there,

if yours only problem is currently the order of fields, inputs, in create/update operation, I think there is no other answer then that, you can't order it anyhow :)

The order of fields, which you create via. `CRUD::field`, `$this->crud->addField([])`, `CRUD::addField([..])` or `CRUD::addFields([..[]..])` or yours `CRUD::field()..` depends only on order, in which you write them in `setupCreateOperation` method (and update method..). So if you have for ex. two fields, and you create them like this:
```
protected function setupCreateOperation()
{
  CRUD::field('country_id')->type('select')->entity('country')->attribute('name')
    ->model('App\Models\Country'); // ->label('Country')
  
  CRUD::field('dealer_name')->type('text')->name('dealer_name')->label('Dealer name');
}
```
Then the order of these fields, inputs, leads to the render in the final form in order, in which you wrote them down in that methods (create, update) which are in the CRUD controller. So every new field which you add to the create/update methods will be at the end of the form.\
In this example the `country_id` would be first, and `dealer_name` field will be at the second position.

There is a way, which I can think of, and that is to create blank fields, and then initialize them right later.. but I highly don't recommend it. I don't think that it is right approach, see below: (I wrote it just to show you the logic behind it)
```
protected function setupCreateOperation()
{
  CRUD::field('country_id');
  CRUD::field('dealer_name');
  
  CRUD::field('dealer_name')->type('text')->name('dealer_name')->label('Dealer name');

  CRUD::field('country_id')->type('select')->entity('country')->attribute('name')
    ->model('App\Models\Country'); // ->label('Country')
}
```
I didn't test it, but from my experience what happened to me, is that, the `country_id` field would be first. Because it was added to `Fields array` if first place. 

Also order is then little bit different, when you use Tabs, .. or, we can say the same (see Backpack Fields API with anchor direct link to this problematics [https://backpackforlaravel.com/docs/6.x/crud-fields#optional-tab-attribute-splits-forms-into-tabs](https://backpackforlaravel.com/docs/6.x/crud-fields#optional-tab-attribute-splits-forms-into-tabs)).\
If you use Tabs, you basically every time you create the field, and specify the tab, Backpack adds it to the end of that tab. Like normally it would add it to the end of a form.

At the end I will just direct you to Backpack documentation which is really rich on examples, together with Backpack demo, links there:
[Backpack docs](https://backpackforlaravel.com/docs/6.x/introduction)
[Backpack demo live app](https://demo.backpackforlaravel.com/admin/login)
[Backpack demo repository on github](https://github.com/laravel-backpack/demo)
From the demo you can take in the repository look at cool examples and usages of Backpack. Developers, thank you for it!

**Comment by yellowbooth** [comment]
Thank you for your reply.
Actually I am using CRUD::setFromDb() which seems to add the country field first for no reason I can see! 
Everything else is added in the order it exists in the table!
You are right, if I add all the fields manually, instead of using setFromDb() - I am able to choose the order, so that does at least allow me a way to solve the issue :-)

**Reply by MaraIGI** [reply]
Heh :D It's one advantage of having English-speaking-future-users of the Backpack CRUD administration, personally I can't use that `setFromDb()` method (due to Column names) and I found out that it was most likely added just recently to Backpack (at least artisan generator commands) - and is cool method for testing, for first CRUD-creation steps, but not for production use IMHO :-)

And also.. with that `setFromDb()` you can't specify `wrapper` attribute in which you can use Bootstrap forms classes logic and reach fancy 4 small fields next to each other rather than display them in 4 rows and etc.

Again, see [attributes and wrapper settings for fields](https://backpackforlaravel.com/docs/6.x/crud-fields#optional-field-attributes-for-presentation-purposes-1) in docs :-)

.. And to answer the problem, the problem, I think, is in that, the relations (specified in Laravel Eloquent Models) from `setFromDb()` are called as first, .. as last, and then/before are called just simple string/int fields etc.

**Reply by yellowbooth** [reply]
Yeah, really I am just being lazy and probably I should define all the fields manually!! To be fair... they have made it pretty user friendly?!
:-)

**Reply by yellowbooth** [reply]
I am getting the hang of this I think.. I just have one further question... currently my country dropdown is defaulting to the first entry... is is possible for the default to be "PLEASE SELECT" or something similar?

**Reply by MaraIGI** [reply]
Is the `country` field in Dealer model nullable? If not, try it to set nullable, then you could also add `placeholder` attribute to that field with "PLEASE SELECT", and if you don't want it to be nullable at the end, still set it to nullable, but add `validation` to that field with `required` rule - but that validation could break your placeholder. I would need to test that to answer more precisely.

**Reply by MaraIGI** [reply]
Ups, completely wrong, .. or, just partly, for select element you can't (??) in HTML add placeholder, you should add option with `value=""` with innerText which you want to display as placeholder. Backpack does this, but adds, to innerText of that default option, just dash. And I don't know if you can't change that dash, unfortunately. But that dash should be displayed when the field in database will be nullable.

**Reply by yellowbooth** [reply]
Thanks MaraIGI - I just found it! :-)
1) I need to make field "required" in DealerRequest.php 
2) I need to specify as ->allows_null(true) in DealerCrudController
This then displays the "-" instead of incorrectly defaulting.
3) I can change the "-" in select.blade.php (although I will probably leave as it is)

Thank you so much for your time.. I appreciate your help :)

**Reply by pxpm** [reply]
Hey @yellowbooth you and @MaraIGI are in the right direction. 
Just an heads up: you are forcing the field to be a `type => select`. That's just the regular HTML select, no placeholder. 

If you use your field as a `select2`, you will have the placeholder option etc that's bundled in select2. https://backpackforlaravel.com/docs/6.x/crud-fields#select2-1-n-relationship-pro

About the ordering of the fields, you can move them where you want using the helper functions: https://backpackforlaravel.com/docs/6.x/crud-api#create-update-operations

Cheers

**Comment by jcastroa87** [comment]
We haven’t heard back from you in a while, so we’re closing this issue for now. If you still need assistance, please don’t hesitate to reopen the issue or create a new one.

Thank you, and take care!

Cheers.
